ClassNotFoundException: com.purbon.topology.roles.RBACProvider

Hi I am getting the following error:

```
kafka-topology-builder.sh --broker localhost:9093 --clientConfig topology.properties --topology topology.yaml 
log4j:WARN No appenders could be found for logger (org.apache.kafka.clients.admin.AdminClientConfig).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Exception in thread "main" java.io.IOException: java.lang.ClassNotFoundException: com.purbon.topology.roles.RBACProvider
        at com.purbon.kafka.topology.KafkaTopologyBuilder.buildAccessControlProvider(KafkaTopologyBuilder.java:208)
        at com.purbon.kafka.topology.KafkaTopologyBuilder.run(KafkaTopologyBuilder.java:78)
        at com.purbon.kafka.topology.BuilderCLI.processTopology(BuilderCLI.java:153)
        at com.purbon.kafka.topology.BuilderCLI.main(BuilderCLI.java:118)
Caused by: java.lang.ClassNotFoundException: com.purbon.topology.roles.RBACProvider
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at com.purbon.kafka.topology.KafkaTopologyBuilder.buildAccessControlProvider(KafkaTopologyBuilder.java:177)
```

topology.yaml
```
team: "team"
source: "source"
projects:
  - name: "bar"
    zookeepers: []
    consumers:
      - principal: "User:Blub"
    producers: []
    streams: []
    connectors: []
    topics:
      - dataType: "json"
        name: "events"
        config:
          replication.factor: "1"
          num.partitions: "1"
    rbac:
      - ResourceOwner:
          - principal: "User:Foo"
```
topology.properties

```
bootstrap.servers=localhost:9093
sasl.mechanism=PLAIN
security.protocol=SASL_PLAINTEXT
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required \
   username="admin" \
   password="admin-secret";

topology.builder.access.control.class="com.purbon.topology.roles.RBACProvider"
topology.builder.mds.server="http://localhost:8090"
topology.builder.mds.user="alice"
topology.builder.mds.password="alice-secret"
topology.builder.mds.kafka.cluster.id="UtBZ3rTSRtypmmkAL1HbHw"
```

When leaving the RBAC-related out of topology properties and yaml file the deployment works like a charm.
Method of deploying was copying descriptor.yaml and topologies file into a docker container running your latest build from dockerhub.

--- a/changelog.md
+++ b/changelog.md
@@ -12,6 +12,7 @@ Note in the future this subsystem should be externalised as plugins
 * raise up error handling in case of non available cluster
 * Make usage of the maven jdeb plugin to build a deb package to use in distributions such as Debian and relative like the Ubuntu family
 * Fix parser problem with non required parameters, such as the ones for the acls where an empty array had to be in place, now all principals for acls if not required can be ignored and will have no effect
+* Ammend param and typo issues with rbac that where not facilitating the execution flow, updated as well the examples to match a correct execution flow
 
 v0.11:
 * Add support for platform wide acls for schema registry in the topology description file.

--- a/example/topology-builder-rbac.properties
+++ b/example/topology-builder-rbac.properties
@@ -1,15 +1,14 @@
-bootstrap.servers=localhost:9092
+bootstrap.servers=localhost:9094
+sasl.mechanism=OAUTHBEARER
 security.protocol=SASL_PLAINTEXT
-sasl.mechanism=PLAIN
-sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required \
-  username="kafka" \
-  password="kafka";
+sasl.login.callback.handler.class=io.confluent.kafka.clients.plugins.auth.token.TokenUserLoginCallbackHandler
+sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required username="professor" password="professor" metadataServerUrls="http://localhost:8090";
 
-topology.builder.access.control.class = "com.purbon.topology.roles.RBACProvider"
-topology.builder.mds.server = "http://localhost:8090"
-topology.builder.mds.user = "mds"
-topology.builder.mds.password = "mds-secret"
-topology.builder.mds.kafka.cluster.id = "foobar"
+topology.builder.access.control.class = com.purbon.kafka.topology.roles.RBACProvider
+topology.builder.mds.server = http://localhost:8090
+topology.builder.mds.user = professor
+topology.builder.mds.password = professor
+topology.builder.mds.kafka.cluster.id = WlDEWYm3TgapNkvVuXT45g
 
-topology.builder.mds.schema.registry.cluster.id = "schema-registry-cluster"
-topology.builder.mds.kafka.connect.cluster.id = "connect-cluster"
+topology.builder.mds.schema.registry.cluster.id = schema-registry-cluster
+topology.builder.mds.kafka.connect.cluster.id = connect-cluster

--- a/example/topology-builder-rbac.properties
+++ b/example/topology-builder-rbac.properties
@@ -8,7 +8,7 @@ topology.builder.access.control.class = com.purbon.kafka.topology.roles.RBACProv
 topology.builder.mds.server = http://localhost:8090
 topology.builder.mds.user = professor
 topology.builder.mds.password = professor
-topology.builder.mds.kafka.cluster.id = WlDEWYm3TgapNkvVuXT45g
+topology.builder.mds.kafka.cluster.id = VFO1W6UsR0yhZCC1h3RtXQ
 
 topology.builder.mds.schema.registry.cluster.id = schema-registry-cluster
 topology.builder.mds.kafka.connect.cluster.id = connect-cluster

--- a/pom.xml
+++ b/pom.xml
@@ -105,6 +105,13 @@
     </profile>
   </profiles>
 
+  <repositories>
+    <repository>
+      <id>confluent</id>
+      <url>https://packages.confluent.io/maven/</url>
+    </repository>
+  </repositories>
+
   <build>
     <finalName>${project.artifactId}</finalName>
     <plugins>
@@ -388,6 +395,8 @@
     <junit.version>4.12</junit.version>
     <testcontainers.version>1.13.0</testcontainers.version>
     <jedis.version>3.2.0</jedis.version>
+    <confluent.version>5.5.1-ce</confluent.version>
+    <jersey.version>2.22.2</jersey.version>
   </properties>
 
   <dependencies>
@@ -408,9 +417,14 @@
       <version>${zookeeper.version}</version>
     </dependency>
     <dependency>
-      <groupId>org.apache.kafka</groupId>
-      <artifactId>kafka-clients</artifactId>
-      <version>${kafka.version}</version>
+      <groupId>io.confluent</groupId>
+      <artifactId>kafka-client-plugins</artifactId>
+      <version>${confluent.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.glassfish.jersey.core</groupId>
+      <artifactId>jersey-common</artifactId>
+      <version>${jersey.version}</version>
     </dependency>
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>

--- a/scripts/run-rbac-with-docker.sh
+++ b/scripts/run-rbac-with-docker.sh
@@ -1,0 +1,11 @@
+#!/usr/bin/env bash
+
+HOME_DIR=`pwd`
+NETWORK="rbac-sasl_default"
+
+
+docker run -v  $HOME_DIR:/app --network $NETWORK \
+    purbon/kafka-topology-builder:latest kafka-topology-builder.sh  \
+    --brokers broker:9092 \
+    --topology /app/example/descriptor-with-rbac.yaml \
+    --clientConfig /app/example/topology-builder-rbac.properties

--- a/src/main/java/com/purbon/kafka/topology/BuilderCLI.java
+++ b/src/main/java/com/purbon/kafka/topology/BuilderCLI.java
@@ -117,6 +117,7 @@ public class BuilderCLI {
     config.put(ADMIN_CLIENT_CONFIG_OPTION, adminClientConfigFile);
     processTopology(topology, config);
     System.out.println("Kafka Topology updated");
+    exit(0);
   }
 
   private static void printHelpOrVersion(

--- a/src/main/java/com/purbon/kafka/topology/KafkaTopologyBuilder.java
+++ b/src/main/java/com/purbon/kafka/topology/KafkaTopologyBuilder.java
@@ -8,6 +8,7 @@ import static com.purbon.kafka.topology.TopologyBuilderConfig.ACCESS_CONTROL_IMP
 import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_KAFKA_CLUSTER_ID_CONFIG;
 import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_KC_CLUSTER_ID_CONFIG;
 import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_PASSWORD_CONFIG;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_SERVER;
 import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_SR_CLUSTER_ID_CONFIG;
 import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_USER_CONFIG;
 import static com.purbon.kafka.topology.TopologyBuilderConfig.RBAC_ACCESS_CONTROL_CLASS;
@@ -82,7 +83,6 @@ public class KafkaTopologyBuilder {
         new AccessControlManager(aclsProvider, cs, cliParams);
 
     TopicManager topicManager = new TopicManager(builderAdminClient, cliParams);
-
     topicManager.sync(topology);
     accessControlManager.sync(topology);
     if (!quiteOut) {
@@ -165,7 +165,7 @@ public class KafkaTopologyBuilder {
     }
   }
 
-  private AccessControlProvider buildAccessControlProvider() throws IOException {
+  public AccessControlProvider buildAccessControlProvider() throws IOException {
     String accessControlClass =
         properties
             .getOrDefault(
@@ -181,7 +181,7 @@ public class KafkaTopologyBuilder {
         return (SimpleAclsProvider) constructor.newInstance(builderAdminClient);
       } else if (accessControlClass.equalsIgnoreCase(RBAC_ACCESS_CONTROL_CLASS)) {
 
-        String mdsServer = properties.getProperty(MDS_USER_CONFIG);
+        String mdsServer = properties.getProperty(MDS_SERVER);
         String mdsUser = properties.getProperty(MDS_USER_CONFIG);
         String mdsPassword = properties.getProperty(MDS_PASSWORD_CONFIG);
 

--- a/src/main/java/com/purbon/kafka/topology/TopologyBuilderConfig.java
+++ b/src/main/java/com/purbon/kafka/topology/TopologyBuilderConfig.java
@@ -8,7 +8,7 @@ public class TopologyBuilderConfig {
   public static final String ACCESS_CONTROL_DEFAULT_CLASS =
       "com.purbon.kafka.topology.roles.SimpleAclsProvider";
   public static final String RBAC_ACCESS_CONTROL_CLASS =
-      "\"com.purbon.kafka.topology.roles.RBACProvider";
+      "com.purbon.kafka.topology.roles.RBACProvider";
 
   public static final String STATE_PROCESSOR_IMPLEMENTATION_CLASS =
       "topology.builder.state.processor.class";

--- a/src/main/java/com/purbon/kafka/topology/roles/RBACProvider.java
+++ b/src/main/java/com/purbon/kafka/topology/roles/RBACProvider.java
@@ -62,14 +62,14 @@ public class RBACProvider implements AccessControlProvider {
     bindings.add(secAdminBinding);
 
     apiClient.bind(principal, DEVELOPER_READ, topicPrefix, PREFIX);
-    if (readTopics != null && readTopics.isEmpty()) {
+    if (readTopics != null && !readTopics.isEmpty()) {
       readTopics.forEach(
           topic -> {
             TopologyAclBinding binding = apiClient.bind(principal, DEVELOPER_READ, topic, LITERAL);
             bindings.add(binding);
           });
     }
-    if (writeTopics != null && readTopics.isEmpty()) {
+    if (writeTopics != null && !writeTopics.isEmpty()) {
       writeTopics.forEach(
           topic -> {
             TopologyAclBinding binding = apiClient.bind(principal, DEVELOPER_WRITE, topic, LITERAL);

--- a/src/test/java/com/purbon/kafka/topology/KafkaTopologyBuilderTest.java
+++ b/src/test/java/com/purbon/kafka/topology/KafkaTopologyBuilderTest.java
@@ -1,6 +1,11 @@
 package com.purbon.kafka.topology;
 
 import static com.purbon.kafka.topology.BuilderCLI.BROKERS_OPTION;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.ACCESS_CONTROL_IMPLEMENTATION_CLASS;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_KAFKA_CLUSTER_ID_CONFIG;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_PASSWORD_CONFIG;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_SERVER;
+import static com.purbon.kafka.topology.TopologyBuilderConfig.MDS_USER_CONFIG;
 import static org.junit.Assert.assertEquals;
 
 import com.purbon.kafka.topology.model.Topology;
@@ -51,5 +56,23 @@ public class KafkaTopologyBuilderTest {
     Topology topology = builder.buildTopology(fileOrDirPath);
 
     assertEquals(4, topology.getProjects().size());
+  }
+
+  @Test
+  public void testRbacSetup() throws URISyntaxException, IOException {
+    URL dirOfDescriptors = getClass().getResource("/dir");
+    String fileOrDirPath = Paths.get(dirOfDescriptors.toURI()).toFile().toString();
+
+    props.put(ACCESS_CONTROL_IMPLEMENTATION_CLASS, "com.purbon.kafka.topology.roles.RBACProvider");
+    props.put(MDS_SERVER, "http://localhost:8090");
+    props.put(MDS_USER_CONFIG, "alice");
+    props.put(MDS_PASSWORD_CONFIG, "alice-secret");
+    props.put(MDS_KAFKA_CLUSTER_ID_CONFIG, "UtBZ3rTSRtypmmkAL1HbHw");
+
+    KafkaTopologyBuilder builder =
+        new KafkaTopologyBuilder(
+            fileOrDirPath, cliOps, new TopologySerdes(), props, topologyAdminClient, false);
+
+    builder.buildAccessControlProvider();
   }
 }

